### ISSUE TITLE
cancelling main starting coroutine firstly (before all other pending tasks)

### ISSUE THREAD

**Issue by FedirAlifirenko** [cjrh/aiorun#42]
Hi, I found the case when I want to cancel main starting coroutine (and await it's cancellation) before the cancellation of all other pending loop's tasks. 
Perhaps, it's not the issue of `aiorun` package directly, but let me ask advice how to deal with such case (if it can be solved using `aiorun`).

The case can be reproduced by using code below:
```python
import argparse
import asyncio
import logging
import signal

from aiorun import run


class SomeThirdPartyRunner:
    def __init__(self):
        self._queue = list(range(5))
        # actually, I can't use `shutdown_waits_for` below because it's 3'rd party package
        self._work_task = asyncio.create_task(self._work())
        self._work_task_done = asyncio.Future()

    async def _work(self):
        try:
            while self._queue:
                logging.info(f'processing queue={self._queue}')
                self._queue.pop()
                await asyncio.sleep(1)
        except asyncio.CancelledError:
            logging.info(' *** we are here only when aiorun is used *** ')

        finally:
            self._work_task_done.set_result(None)

    async def wait_done(self):
        await self._work_task_done


async def corofn():
    runner = SomeThirdPartyRunner()
    try:
        await asyncio.sleep(2)

    except asyncio.CancelledError:
        pass

    finally:
        logging.info('stopping runner...')
        await runner.wait_done()
        logging.info('runner stopped')


def run_through_aiorun():
    run(_aiorun_main())


async def _aiorun_main():
    await corofn()


def run_through_asyncio():
    asyncio.run(_asyncio_main())


async def _asyncio_main():
    loop = asyncio.get_event_loop()

    task = loop.create_task(corofn())
    task.add_done_callback(lambda _: loop.stop())

    loop.add_signal_handler(signal.SIGINT, task.cancel)
    loop.add_signal_handler(signal.SIGTERM, task.cancel)

    await task


if __name__ == '__main__':
    logging.basicConfig(
        level=logging.DEBUG,
        format="%(asctime)s - [%(levelname)s] - [%(name)s] "
               "- %(filename)s:%(lineno)d - %(message)s",
    )

    parser = argparse.ArgumentParser()
    parser.add_argument('--aiorun',
                        action='store_true',
                        help='Run script using aiorun package')
    args = parser.parse_args()

    if args.aiorun:
        logging.info('Running using aiorun')
        run_through_aiorun()
    else:
        logging.info('Running using asyncio')
        run_through_asyncio()

```

Starting script using `aiorun` and press Ctrl+C on 2'nd message `processing queue`:
```text
(env_3_8) MacBook-Pro-2:~ fedir$ python script.py --aiorun
[INFO] - [root] - script.py:83 - Running using aiorun
[DEBUG] - [aiorun] - aiorun.py:155 - Entering run()
[DEBUG] - [asyncio] - selector_events.py:59 - Using selector: KqueueSelector
[DEBUG] - [aiorun] - aiorun.py:236 - Creating default executor
[INFO] - [root] - script.py:18 - processing queue=[0, 1, 2, 3, 4]
[INFO] - [root] - script.py:18 - processing queue=[0, 1, 2, 3]
^C[DEBUG] - [aiorun] - aiorun.py:304 - Entering shutdown handler
[CRITICAL] - [aiorun] - aiorun.py:317 - Stopping the loop
[INFO] - [aiorun] - aiorun.py:249 - Entering shutdown phase.
[INFO] - [aiorun] - aiorun.py:262 - Cancelling pending tasks.
[DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-1' coro=<run.<locals>.new_coro() running at /Users/fedir/env/env_3_8/lib/python3.8/site-packages/aiorun.py:206> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x10a00fdf0>()]>>
[DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-2' coro=<SomeThirdPartyRunner._work() running at script.py:20> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x10a00fe50>()]>>
[INFO] - [aiorun] - aiorun.py:276 - Running pending tasks till complete
[INFO] - [root] - script.py:40 - stopping runner...
[INFO] - [root] - script.py:22 -  *** we are here only when aiorun is used ***
[INFO] - [root] - script.py:42 - runner stopped
[INFO] - [aiorun] - aiorun.py:281 - Waiting for executor shutdown.
[INFO] - [aiorun] - aiorun.py:286 - Shutting down async generators
[INFO] - [aiorun] - aiorun.py:288 - Closing the loop.
[INFO] - [aiorun] - aiorun.py:290 - Leaving. Bye!
```
**Desired** result for me in this case is do not cancel  `corofn()` along with `SomeThirdPartyRunner._work()`.

Starting script using `asyncio` and press Ctrl+C on the same log msg:
```text
(env_3_8) MacBook-Pro-2:~ fedir$ python script.py
[INFO] - [root] - script.py:86 - Running using asyncio
[DEBUG] - [asyncio] - selector_events.py:59 - Using selector: KqueueSelector
[INFO] - [root] - script.py:18 - processing queue=[0, 1, 2, 3, 4]
[INFO] - [root] - script.py:18 - processing queue=[0, 1, 2, 3]
^C[INFO] - [root] - script.py:40 - stopping runner...
[INFO] - [root] - script.py:18 - processing queue=[0, 1, 2]
[INFO] - [root] - script.py:18 - processing queue=[0, 1]
[INFO] - [root] - script.py:18 - processing queue=[0]
[INFO] - [root] - script.py:42 - runner stopped
```

I've done some updates into `aiorun` to receive expected result:
https://github.com/cjrh/aiorun/blob/master/aiorun.py#L210
```python
_origin_coro_task = loop.create_task(new_coro())
```
and then on the start of "Entering the shutdown phase":
https://github.com/cjrh/aiorun/blob/master/aiorun.py#L249
```python
if _origin_coro_task is not None:
    logger.debug("Cancelling origin coro task: %s", _origin_coro_task)
    _origin_coro_task.cancel()
    loop.run_until_complete(_origin_coro_task)
```

And received expected result below:
```text
(env_3_8) MacBook-Pro-2:~ fedir$ python script.py --aiorun
[INFO] - [root] - script.py:83 - Running using aiorun
[DEBUG] - [aiorun] - aiorun.py:155 - Entering run()
[DEBUG] - [asyncio] - selector_events.py:59 - Using selector: KqueueSelector
[DEBUG] - [aiorun] - aiorun.py:237 - Creating default executor
[INFO] - [root] - script.py:18 - processing queue=[0, 1, 2, 3, 4]
[INFO] - [root] - script.py:18 - processing queue=[0, 1, 2, 3]
^C[DEBUG] - [aiorun] - aiorun.py:309 - Entering shutdown handler
[CRITICAL] - [aiorun] - aiorun.py:322 - Stopping the loop
[INFO] - [aiorun] - aiorun.py:250 - Entering shutdown phase.
[DEBUG] - [aiorun] - aiorun.py:252 - Cancelling origin coro task: <Task pending name='Task-1' coro=<run.<locals>.new_coro() running at /Users/fedir/env/env_3_8/lib/python3.8/site-packages/aiorun.py:207> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x105197d90>()]>>
[INFO] - [root] - script.py:40 - stopping runner...
[INFO] - [root] - script.py:18 - processing queue=[0, 1, 2]
[INFO] - [root] - script.py:18 - processing queue=[0, 1]
[INFO] - [root] - script.py:18 - processing queue=[0]
[INFO] - [root] - script.py:42 - runner stopped
[INFO] - [aiorun] - aiorun.py:267 - Cancelling pending tasks.
[INFO] - [aiorun] - aiorun.py:281 - Running pending tasks till complete
[INFO] - [aiorun] - aiorun.py:286 - Waiting for executor shutdown.
[INFO] - [aiorun] - aiorun.py:291 - Shutting down async generators
[INFO] - [aiorun] - aiorun.py:293 - Closing the loop.
[INFO] - [aiorun] - aiorun.py:295 - Leaving. Bye!
```

It would be great to receive expected result using the current aiorun functionality. Thx in advance for feedback.
(env versions)
```text
(env_3_8) MacBook-Pro-2:~ fedir$ python -V
Python 3.8.1
(env_3_8) MacBook-Pro-2:~ fedir$ pip freeze | grep aiorun
aiorun==2020.1.3
```

P.S. I can provide more practical example of such behaviour if needed (using the combination of `aiorun` and `aiokafka` packages).

**Comment by cjrh** [comment]
@FedirAlifirenko Hi, and thanks for the report. It may take me a few days to get to it but it looks interesting.  I just wanted to let you know I've seen your report and I will definitely go through it as soon as I have time available.

**Comment by cjrh** [comment]
I had a look at the example.  I _think_ I understand your question, but if not I apologise.  Anyhow: the processing of queues is a very common situation where we are unable to cancel tasks immediately - because there is work on the queue that we don't want to drop!  In the most general case, you really want to use a proper task processing like celery, luigi or huey because these tools save their pending queued work in a separate persistent store, which means their services can be restarted easily between each individual job. They pick up the remaining work after restart.

Nevertheless, often celery is overkill for ad hoc situations and smaller workloads.  In asyncio there are two main strategies for allowing a queue to be drained during the shutdown sequence:

1) Absorb `CancelledError`, and keep going until you get a `None` on the queue
2) Don't cancel the queue-processing task at all.

## 1) Absorbing cancellation

The first one, in your code, would have the worker coroutine function look something like this:

```
    async def _work(self):
        while self._queue:
            logging.info(f'processing queue={self._queue}')
            self._queue.pop()
            try:
                await asyncio.sleep(1)
            except asyncio.CancelledError:
                continue

        self._work_task_done.set_result(None)
```

Running the above produces the expected output I think (running on docker image python:3.8.1-slim-buster)?

```
root@ab5c261b5c18:/src# python main-simpler.py --aiorun
2020-02-02 00:17:25,434 - [INFO] - [root] - main-simpler.py:85 - Running using aiorun
2020-02-02 00:17:25,434 - [DEBUG] - [aiorun] - aiorun.py:155 - Entering run()
2020-02-02 00:17:25,434 - [DEBUG] - [asyncio] - selector_events.py:59 - Using selector: EpollSelector
2020-02-02 00:17:25,435 - [DEBUG] - [aiorun] - aiorun.py:236 - Creating default executor
2020-02-02 00:17:25,436 - [INFO] - [root] - main-simpler.py:20 - processing queue=[0, 1, 2, 3, 4]
2020-02-02 00:17:26,437 - [INFO] - [root] - main-simpler.py:20 - processing queue=[0, 1, 2, 3]
^C2020-02-02 00:17:27,013 - [DEBUG] - [aiorun] - aiorun.py:304 - Entering shutdown handler
2020-02-02 00:17:27,014 - [CRITICAL] - [aiorun] - aiorun.py:317 - Stopping the loop
2020-02-02 00:17:27,014 - [INFO] - [aiorun] - aiorun.py:249 - Entering shutdown phase.
2020-02-02 00:17:27,014 - [INFO] - [aiorun] - aiorun.py:262 - Cancelling pending tasks.
2020-02-02 00:17:27,014 - [DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-3' coro=<shutdown_waits_for.<loc
als>.inner() running at /usr/local/lib/python3.8/site-packages/aiorun.py:110> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object
at 0x7f28f3da68b0>()]>>
2020-02-02 00:17:27,014 - [DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-1' coro=<run.<locals>.new_coro()
 running at /usr/local/lib/python3.8/site-packages/aiorun.py:206> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f28f3d
a6850>()]>>
2020-02-02 00:17:27,015 - [INFO] - [aiorun] - aiorun.py:276 - Running pending tasks till complete
2020-02-02 00:17:27,015 - [INFO] - [root] - main-simpler.py:42 - stopping runner...
2020-02-02 00:17:27,439 - [INFO] - [root] - main-simpler.py:20 - processing queue=[0, 1, 2]
2020-02-02 00:17:28,440 - [INFO] - [root] - main-simpler.py:20 - processing queue=[0, 1]
2020-02-02 00:17:29,442 - [INFO] - [root] - main-simpler.py:20 - processing queue=[0]
2020-02-02 00:17:30,444 - [INFO] - [root] - main-simpler.py:44 - runner stopped
2020-02-02 00:17:30,444 - [INFO] - [aiorun] - aiorun.py:281 - Waiting for executor shutdown.
2020-02-02 00:17:30,444 - [INFO] - [aiorun] - aiorun.py:286 - Shutting down async generators
2020-02-02 00:17:30,444 - [INFO] - [aiorun] - aiorun.py:288 - Closing the loop.
2020-02-02 00:17:30,446 - [INFO] - [aiorun] - aiorun.py:290 - Leaving. Bye!
```

To absorb cancellation, the key thing is to handle all the `await` sections, because the `CancelledError` exception can be raised at any of them. Your work function only has one.  Your code atypical of a queue processing task for a number of reasons: you're not using `asyncio.Queue`, for example, and you're not waiting on a sentinel value on the queue (like `None`) to exit.

The more typical setup here would be to use `asyncio.Queue`, and feed it a `None` when no more work will be coming in. Then, the queue-processing function will absorb cancellation, but will exit when it receives `None` on the queue.  This way you can make sure that all the pending work on the queue will still get processed.  A more conventional setup might look something like this:

(I'm modifying your code to illustrate)

```
class SomeThirdPartyRunner:
    def __init__(self, q: asyncio.Queue):
        self._queue = q
        self._work_task = asyncio.create_task(self._work())
        self._work_task_done = asyncio.Future()

    async def _work(self):
        while True:
            try:
                logging.info(f'processing queue={self._queue}')
                work = await self._queue.get()
            except asyncio.CancelledError:
                # Ignore cancellation, just keep going
                continue

            if work is None:
                # This should be the only way to exit from this function.
                break

            try:
                # Do work
                await asyncio.sleep(1)
                self._queue.task_done()
            except asyncio.CancelledError:
                # Ignore cancellation and start the work again. This is
                # actually quite tricky to manage in a general case, i.e.,
                # "restarting" cancelled work - depends on specific use case.
                await asyncio.sleep(1)

        self._work_task_done.set_result(None)

    async def wait_done(self):
        await self._work_task_done


async def corofn():
    q = asyncio.Queue()
    runner = SomeThirdPartyRunner(q)

    # Note how work can be added to the queue after the worker has
    # already been set up
    for i in range(5):
        await q.put(i)

    try:
        await runner.wait_done()
    except asyncio.CancelledError:
        logging.info('stopping runner...')
        await q.put(None)
    finally:
        await runner.wait_done()
        logging.info('runner stopped')
```

All the other parts of the file are the same. This ^ is a more typical setup where the worker job waits only for `None` on the queue before exiting.   It shuts down correctly either before or after all the queue jobs are complete:

```
CTRL-C after all the work is already complete:

root@ba840b1626fa:/src# python main-conventional.py --aiorun
2020-02-02 00:45:47,451 - [INFO] - [root] - main-conventional.py:105 - Running using aiorun
2020-02-02 00:45:47,451 - [DEBUG] - [aiorun] - aiorun.py:155 - Entering run()
2020-02-02 00:45:47,451 - [DEBUG] - [asyncio] - selector_events.py:59 - Using selector: EpollSelector
2020-02-02 00:45:47,452 - [DEBUG] - [aiorun] - aiorun.py:236 - Creating default executor
2020-02-02 00:45:47,452 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[0, 1, 2, 3, 4] tasks=5>
2020-02-02 00:45:48,454 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[1, 2, 3, 4] tasks=4>
2020-02-02 00:45:49,455 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[2, 3, 4] tasks=3>
2020-02-02 00:45:50,457 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[3, 4] tasks=2>
2020-02-02 00:45:51,458 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[4] tasks=1>
2020-02-02 00:45:52,460 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0>
^C2020-02-02 00:45:55,974 - [DEBUG] - [aiorun] - aiorun.py:304 - Entering shutdown handler
2020-02-02 00:45:55,975 - [CRITICAL] - [aiorun] - aiorun.py:317 - Stopping the loop
2020-02-02 00:45:55,975 - [INFO] - [aiorun] - aiorun.py:249 - Entering shutdown phase.
2020-02-02 00:45:55,975 - [INFO] - [aiorun] - aiorun.py:262 - Cancelling pending tasks.
2020-02-02 00:45:55,975 - [DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-2' coro=<SomeThirdPartyRunner._w
ork() running at main-conventional.py:20> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fbabc246790>()]>>
2020-02-02 00:45:55,975 - [DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-1' coro=<run.<locals>.new_coro()
 running at /usr/local/lib/python3.8/site-packages/aiorun.py:206> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fbabc2
467c0>()]>>
2020-02-02 00:45:55,976 - [INFO] - [aiorun] - aiorun.py:276 - Running pending tasks till complete
2020-02-02 00:45:55,976 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0>
2020-02-02 00:45:55,976 - [INFO] - [root] - main-conventional.py:57 - stopping runner...
2020-02-02 00:45:55,976 - [INFO] - [aiorun] - aiorun.py:281 - Waiting for executor shutdown.
2020-02-02 00:45:55,976 - [INFO] - [aiorun] - aiorun.py:286 - Shutting down async generators
2020-02-02 00:45:55,976 - [INFO] - [aiorun] - aiorun.py:288 - Closing the loop.
2020-02-02 00:45:55,978 - [INFO] - [aiorun] - aiorun.py:290 - Leaving. Bye!

CTRL-C before all the work is complete:

root@ba840b1626fa:/src# python main-conventional.py --aiorun
2020-02-02 00:45:58,341 - [INFO] - [root] - main-conventional.py:105 - Running using aiorun
2020-02-02 00:45:58,341 - [DEBUG] - [aiorun] - aiorun.py:155 - Entering run()
2020-02-02 00:45:58,341 - [DEBUG] - [asyncio] - selector_events.py:59 - Using selector: EpollSelector
2020-02-02 00:45:58,343 - [DEBUG] - [aiorun] - aiorun.py:236 - Creating default executor
2020-02-02 00:45:58,343 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[0, 1, 2, 3, 4] tasks=5>
2020-02-02 00:45:59,345 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[1, 2, 3, 4] tasks=4>
2020-02-02 00:46:00,347 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[2, 3, 4] tasks=3>
^C2020-02-02 00:46:01,197 - [DEBUG] - [aiorun] - aiorun.py:304 - Entering shutdown handler
2020-02-02 00:46:01,199 - [CRITICAL] - [aiorun] - aiorun.py:317 - Stopping the loop
2020-02-02 00:46:01,199 - [INFO] - [aiorun] - aiorun.py:249 - Entering shutdown phase.
2020-02-02 00:46:01,200 - [INFO] - [aiorun] - aiorun.py:262 - Cancelling pending tasks.
2020-02-02 00:46:01,200 - [DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-2' coro=<SomeThirdPartyRunner._w
ork() running at main-conventional.py:31> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f13ad4e2790>()]>>
2020-02-02 00:46:01,200 - [DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-1' coro=<run.<locals>.new_coro()
 running at /usr/local/lib/python3.8/site-packages/aiorun.py:206> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f13ad4
e27c0>()]>>
2020-02-02 00:46:01,201 - [INFO] - [aiorun] - aiorun.py:276 - Running pending tasks till complete
2020-02-02 00:46:01,201 - [INFO] - [root] - main-conventional.py:57 - stopping runner...
2020-02-02 00:46:02,203 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[3, 4, None] tasks=4>
2020-02-02 00:46:03,204 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[4, None] tasks=3>
2020-02-02 00:46:04,206 - [INFO] - [root] - main-conventional.py:19 - processing queue=<Queue maxsize=0 _queue=[None] tasks=2>
2020-02-02 00:46:04,206 - [INFO] - [aiorun] - aiorun.py:281 - Waiting for executor shutdown.
2020-02-02 00:46:04,206 - [INFO] - [aiorun] - aiorun.py:286 - Shutting down async generators
2020-02-02 00:46:04,207 - [INFO] - [aiorun] - aiorun.py:288 - Closing the loop.
2020-02-02 00:46:04,209 - [INFO] - [aiorun] - aiorun.py:290 - Leaving. Bye!
```

## 2) Don't cancel the queue-processing task at all.

In this case, using `shutdown_waits_for()` is the simplest answer, but you said that you can't use it because the creation of the worker task is in a 3rd-party library and you don't have access to it.

For reference it would look like this:

```
class SomeThirdPartyRunner:
    def __init__(self):
        self._queue = list(range(5))
        # actually, I can't use `shutdown_waits_for` below because it's 3'rd party package
        self._work_task = asyncio.create_task(
            shutdown_waits_for(self._work())
        )
        self._work_task_done = asyncio.Future()
```

And that would be all you need to do (I know you can't because of the 3rd party issue, but just wanted to show this.).

Does this help?

**Comment by cjrh** [comment]
I was confused why the standard library `asyncio.run` was allowing the `_work` coroutine to proceed, but now I see that this is because you're adding your own custom signal handler which doesn't do the default "cancel all tasks" operation.

You can do exactly the same with aiorun, replace only this function:

```
def run_through_aiorun():
    loop = asyncio.get_event_loop()

    task = loop.create_task(corofn())
    task.add_done_callback(lambda _: loop.stop())

    loop.add_signal_handler(signal.SIGINT, task.cancel)
    loop.add_signal_handler(signal.SIGTERM, task.cancel)
    run(loop=loop)
```

This produces the same output as asyncio.run:

```
root@ba840b1626fa:/src# python main-loop.py --aiorun
2020-02-02 01:05:48,164 - [INFO] - [root] - main-loop.py:89 - Running using aiorun
2020-02-02 01:05:48,164 - [DEBUG] - [asyncio] - selector_events.py:59 - Using selector: EpollSelector
2020-02-02 01:05:48,165 - [DEBUG] - [aiorun] - aiorun.py:155 - Entering run()
2020-02-02 01:05:48,166 - [DEBUG] - [aiorun] - aiorun.py:236 - Creating default executor
2020-02-02 01:05:48,166 - [INFO] - [root] - main-loop.py:21 - processing queue=[0, 1, 2, 3, 4]
2020-02-02 01:05:49,168 - [INFO] - [root] - main-loop.py:21 - processing queue=[0, 1, 2, 3]
^C2020-02-02 01:05:49,910 - [DEBUG] - [aiorun] - aiorun.py:304 - Entering shutdown handler
2020-02-02 01:05:49,913 - [CRITICAL] - [aiorun] - aiorun.py:317 - Stopping the loop
2020-02-02 01:05:49,913 - [INFO] - [aiorun] - aiorun.py:249 - Entering shutdown phase.
2020-02-02 01:05:49,913 - [INFO] - [aiorun] - aiorun.py:262 - Cancelling pending tasks.
2020-02-02 01:05:49,913 - [DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-1' coro=<corofn() running at mai
n-loop.py:37> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fa7009e3070>()]> cb=[run_through_aiorun.<locals>.<lambda>(
) at main-loop.py:52]>
2020-02-02 01:05:49,914 - [DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-3' coro=<shutdown_waits_for.<loc
als>.inner() running at /usr/local/lib/python3.8/site-packages/aiorun.py:110> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object
at 0x7fa7009e30d0>()]>>
2020-02-02 01:05:49,914 - [INFO] - [aiorun] - aiorun.py:276 - Running pending tasks till complete
2020-02-02 01:05:49,914 - [INFO] - [root] - main-loop.py:43 - stopping runner...
2020-02-02 01:05:50,169 - [INFO] - [root] - main-loop.py:21 - processing queue=[0, 1, 2]
2020-02-02 01:05:51,171 - [INFO] - [root] - main-loop.py:21 - processing queue=[0, 1]
2020-02-02 01:05:52,172 - [INFO] - [root] - main-loop.py:21 - processing queue=[0]
2020-02-02 01:05:53,174 - [INFO] - [root] - main-loop.py:45 - runner stopped
2020-02-02 01:05:53,174 - [INFO] - [aiorun] - aiorun.py:281 - Waiting for executor shutdown.
2020-02-02 01:05:53,174 - [INFO] - [aiorun] - aiorun.py:290 - Leaving. Bye!
```

Is there much benefit to using `aiorun` over `asyncio.run` for your use-case?  That's difficult for me to say.  There might be other ways to structure your code that will work better with the 3rd party modules, but it's difficult for me to say.

**Comment by FedirAlifirenko** [comment]
Thanks for the detailed answers and proposals. Maybe I didn't fully reveal the main question, I'll try to clarify it below after comments.

**1) Absorbing cancellation**
Yes, handling the cancellation inside the `while` block in the `_work` coroutine gives the finalization of queue processing. It's ok, but I'm expecting another result ().
**2) Don't cancel the queue-processing task at all.**
Yes, using `shutdown_waits_for` gives the same behaviour (`_work` finished after queue was processed).

> Your work function only has one. Your code atypical of a queue processing task for a number of reasons: you're not using asyncio.Queue, for example, and you're not waiting on a sentinel value on the queue (like None) to exit.

I didn't use it to make example more simpler (anyway thanks for these advices).

**3) Question clarification:**
The question is **how to cancel `corofn()` and await it's cancellation before starting the cancellation of `_work()` and other loop's tasks?** (of course using aiorun).

Let me show you one more example below to understand the question's context better.

The example needs kafka to be started as docker container, the content of `docker-compose.yml` is below:
```
version: '3.2'

services:
  kafka:
    image: wurstmeister/kafka
    ports:
      - 9092:9092
    depends_on:
      - zookeeper
    environment:
      KAFKA_ADVERTISED_HOST_NAME: localhost
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_CREATE_TOPICS: "items:2:1"
      KAFKA_AUTO_CREATE_TOPICS_ENABLE: "true"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock

  zookeeper:
    image: wurstmeister/zookeeper
    ports:
      - 2181:2181
```
Starting kafka using `docker-compose up kafka`.

Simple consumer script using `aiokafka` and `aiorun` as entrypoint:
```python
import asyncio
import logging

from aiokafka import AIOKafkaConsumer
from aiorun import run, shutdown_waits_for

logger = logging.getLogger(__name__)


async def corofn():
    consumer = AIOKafkaConsumer(
        'items',
        loop=asyncio.get_event_loop(),
        bootstrap_servers='localhost:9092',
        group_id='test-items-group',
        request_timeout_ms=5000,  # decreased default value 40s only for quicker finish
    )
    await consumer.start()
    logger.info('consumer started...')
    try:
        async for msg in consumer:
            print(msg)

    except asyncio.CancelledError:
        pass

    finally:
        logger.info('stopping kafka consumer...')
        await shutdown_waits_for(consumer.stop())
        logger.info('stopped kafka consumer')


async def main():
    await corofn()


if __name__ == '__main__':
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s - [%(levelname)s] - [%(name)s] - %(filename)s:%(lineno)d - %(message)s",
    )
    run(main())
```

Press `Ctrl+C` after `consumer started...` msg (no message processing or producing, simple start & stop is enough for the example):
```text
(env_3_8) MacBook-Pro-2:~ fedir$ python aiorun_aiokafka.py 
2020-02-03 10:55:09,471 - [INFO] - [aiokafka.consumer.subscription_state] - subscription_state.py:107 - Updating subscribed topics to: frozenset({'items'})
2020-02-03 10:55:09,513 - [INFO] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:548 - Discovered coordinator 1001 for group test-items-group
2020-02-03 10:55:09,513 - [INFO] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:383 - Revoking previously assigned partitions set() for group test-items-group
2020-02-03 10:55:09,513 - [INFO] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:1204 - (Re-)joining group test-items-group
2020-02-03 10:55:09,521 - [INFO] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:1259 - Joined group 'test-items-group' (generation 50) with member_id aiokafka-0.5.2-e5bf6586-e1c0-474d-a94c-25a74ca19ecb
2020-02-03 10:55:09,521 - [INFO] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:1263 - Elected group leader -- performing partition assignments using roundrobin
2020-02-03 10:55:09,528 - [INFO] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:1380 - Successfully synced group test-items-group with generation 50
2020-02-03 10:55:09,528 - [INFO] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:460 - Setting newly assigned partitions {TopicPartition(topic='items', partition=0)} for group test-items-group
2020-02-03 10:55:09,528 - [INFO] - [__main__] - aiorun_aiokafka.py:19 - consumer started...
^C2020-02-03 10:55:12,018 - [CRITICAL] - [aiorun] - aiorun.py:317 - Stopping the loop
2020-02-03 10:55:12,018 - [INFO] - [aiorun] - aiorun.py:249 - Entering shutdown phase.
2020-02-03 10:55:12,019 - [INFO] - [aiorun] - aiorun.py:262 - Cancelling pending tasks.
2020-02-03 10:55:12,019 - [INFO] - [aiorun] - aiorun.py:276 - Running pending tasks till complete
2020-02-03 10:55:12,019 - [INFO] - [__main__] - aiorun_aiokafka.py:28 - stopping kafka consumer...
2020-02-03 10:55:17,024 - [ERROR] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:283 - Error sending LeaveGroupRequest_v1 to node 1001 [[Error 7] RequestTimedOutError] -- marking coordinator dead
2020-02-03 10:55:17,024 - [WARNING] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:482 - Marking the coordinator dead (node 1001)for group test-items-group.
2020-02-03 10:55:17,024 - [ERROR] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:355 - LeaveGroup request failed: [Error 7] RequestTimedOutError
2020-02-03 10:55:17,024 - [INFO] - [__main__] - aiorun_aiokafka.py:30 - stopped kafka consumer
2020-02-03 10:55:17,025 - [INFO] - [aiorun] - aiorun.py:281 - Waiting for executor shutdown.
2020-02-03 10:55:17,025 - [INFO] - [aiorun] - aiorun.py:286 - Shutting down async generators
2020-02-03 10:55:17,025 - [INFO] - [aiorun] - aiorun.py:288 - Closing the loop.
2020-02-03 10:55:17,026 - [INFO] - [aiorun] - aiorun.py:290 - Leaving. Bye!
```

And we are receiving errors on consumer stopping phase (when consumer sending leave group request `Error sending LeaveGroupRequest_v1 to node 1001`).

If you'll change the log_level to `DEBUG` you'll see logs something like that:
```
[DEBUG] - [aiorun] - aiorun.py:264 - Cancelling task: <Task pending name='Task-11' coro=<AIOKafkaConnection._read() running at /Users/fedir/env/env_3_8/lib/python3.8/site-packages/aiokafka/conn.py:475> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x10d970a60>()]> cb=[AIOKafkaConnection._on_read_task_error(<weakref at 0...t 0x10d9633d0>)() at /Users/fedir/env/env_3_8/lib/python3.8/site-packages/aiokafka/conn.py:331]>
```
As I see, from logs and source code, the reason of `LeaveGroupRequest_v1` error is that  `AIOKafkaConnection._read()` task was cancelled along with `corofn()` here https://github.com/cjrh/aiorun/blob/master/aiorun.py#L265 

Yeah, maybe it's also the point to the `aiokafka` internals but handling the order of cancellation (firstly `corofn()` and then all others tasks) helps me to receive expected successfull consumer stopping.

P.S. Again, it's not the `aiorun` functionality issue it's like the request how to deal with such cases using `aiorun` package.

**Comment by cjrh** [comment]
I have reproduced what you observe, using the docker-compose file and your example.

- In your try-finally (after `CancelledError` has been raised inside all the running tasks), we call `consumer.stop()`, same as in the aiokafka docs (I read through their guide).
- Inside `consumer.stop`, aiokafka calls `self._coordinator.close()`:

```
    @asyncio.coroutine
    def stop(self):
        """ Close the consumer, while waiting for finilizers:

            * Commit last consumed message if autocommit enabled
            * Leave group if used Consumer Groups
        """
        if self._closed:
            return
        log.debug("Closing the KafkaConsumer.")
        self._closed = True
        if self._coordinator:
            yield from self._coordinator.close()
        if self._fetcher:
            yield from self._fetcher.close()
        yield from self._client.close()
        log.debug("The KafkaConsumer has closed.")
```
- Inside the group coordinator close, aiokafka calls `self._maybe_leave_group()`:
```
    @asyncio.coroutine
    def close(self):
        """Close the coordinator, leave the current group
        and reset local generation/memberId."""
        if self._closing.done():
            return

        self._closing.set_result(None)
        # We must let the coordination task properly finish all pending work
        if not self._coordination_task.done():
            yield from self._coordination_task
        yield from self._stop_heartbeat_task()
        yield from self._stop_commit_offsets_refresh_task()

        yield from self._maybe_leave_group()     # <---- HERE
```
- Inside `_maybe_leave_group`, aiokafka calls `self._send_req(request)`, to leave the group:
```
    @asyncio.coroutine
    def _maybe_leave_group(self):
        if self.generation > 0:
            # this is a minimal effort attempt to leave the group. we do not
            # attempt any resending if the request fails or times out.
            version = 0 if self._client.api_version < (0, 11, 0) else 1
            request = LeaveGroupRequest[version](self.group_id, self.member_id)
            try:
                yield from self._send_req(request)
            except Errors.KafkaError as err:
                log.error("LeaveGroup request failed: %s", err)
            else:
                log.info("LeaveGroup request succeeded")
        self.reset_generation()
```
- It never gets a response, because the `_read` task has already been killed (as you already figured out)
- The `_read` task just blindly absorbs cancellation in its `add_done_callback()`:
```
    @staticmethod
    def _on_read_task_error(self_ref, read_task):
        # We don't want to react to cancelled errors
        if read_task.cancelled():
            print('read_task was cancelled')   
            return                             # <----- HERE

        try:
            read_task.result()
        except (OSError, EOFError, ConnectionError) as exc:
            self_ref().close(reason=CloseReason.CONNECTION_BROKEN, exc=exc)
        except Exception as exc:
            self = self_ref()
            self.log.exception("Unexpected exception in AIOKafkaConnection")
            self.close(reason=CloseReason.CONNECTION_BROKEN, exc=exc)
```
- I added the `print()` in the snipped above, you can see it print out after you press CTRL-C

This problem is not specific to aiorun: the same problem happens with the standard library `asyncio.run`, unless you override the default cancel-all-tasks behaviour by implementing your own signal handler, as you've already shown.

So I think in the short-to-medium term your best option is to set up a custom signal handler, like you've already done, and control the shutdown sequence.

It is not good for libraries to require special considerations for shutdown - if many libraries did this, it would become very difficult to combine several of them together in one asyncio service. This is why all libraries should be designed so that _all_ tasks can receive `CancelledError` and shut down in an orderly process. Anyway, that's not up to you and me I guess.

aiokafka will have to change how their shutdown process works in order to be compatible with default asyncio.run behaviour (and also therefore `aiorun`). It's telling that the aiokafka examples only use `loop.run_until_complete()` instead of `asyncio.run`.
I had a look at the aiokafka source code. What they could do is absorb cancellation inside the `AIOKafkaConnection._read` method. This the current `_read` method:

```
    @classmethod
    @asyncio.coroutine
    def _read(cls, self_ref):
        # XXX: I know that it become a bit more ugly once cyclic references
        # were removed, but it's needed to allow connections to properly
        # release resources if leaked.
        # NOTE: all errors will be handled by done callback

        reader = self_ref()._reader
        while True:
            resp = yield from reader.readexactly(4)
            size, = struct.unpack(">i", resp)

            resp = yield from reader.readexactly(size)
            self_ref()._handle_frame(resp)
```

Perhaps something like this might work:

```
    @classmethod
    @asyncio.coroutine
    def _read(cls, self_ref):
        # XXX: I know that it become a bit more ugly once cyclic references
        # were removed, but it's needed to allow connections to properly
        # release resources if leaked.
        # NOTE: all errors will be handled by done callback

        reader = self_ref()._reader
        while True:
            try:
                resp = yield from reader.readexactly(4)
            except asyncio.CancelledError:
                continue
            size, = struct.unpack(">i", resp)

            resp = yield from reader.readexactly(size)
            self_ref()._handle_frame(resp)
```

I did only very limited testing, but this appears to shutdown clean. (It's still qutie messy though).

**Comment by FedirAlifirenko** [comment]
>I have reproduced what you observe, using the docker-compose file and your example.

thx for reproducing and aiokafka's internals report  ✅ 

> This problem is not specific to aiorun: the same problem happens with the standard library asyncio.run, unless you override the default cancel-all-tasks behaviour by implementing your own signal handler, as you've already shown.
I don't think so. It seems it's not the same problem because the standard `asyncio.run` implementation doesn't handle signals (how aiorun does it). And when I add signals handling for `asyncio.run` it helps me to not receive leave group request error.

```python
import argparse
import asyncio
import logging
import signal

import aiorun
from aiokafka import AIOKafkaConsumer

logger = logging.getLogger(__name__)
args = None


async def corofn():
    consumer = AIOKafkaConsumer(
        'items',
        loop=asyncio.get_event_loop(),
        bootstrap_servers='localhost:9092',
        group_id='test-items-group',
        request_timeout_ms=5000,  # decreased default value 40s only for quicker finish
    )
    await consumer.start()
    logger.info('consumer started...')
    try:
        async for _ in consumer:
            pass

    except asyncio.CancelledError:
        pass

    finally:
        logger.info('stopping kafka consumer...')

        def _log_stopped_cb(_):
            logger.info('stopped kafka consumer')

        if args.aiorun:
            stop_t = asyncio.create_task(aiorun.shutdown_waits_for(consumer.stop()))
            stop_t.add_done_callback(_log_stopped_cb)
            await stop_t
        else:
            stop_t = asyncio.create_task(consumer.stop())
            stop_t.add_done_callback(_log_stopped_cb)
            await asyncio.shield(stop_t)


def run(use_aiorun: bool):
    if use_aiorun:
        logging.info('Running using aiorun')
        aiorun.run(_aiorun_main())
    else:
        logging.info('Running using asyncio')
        asyncio.run(_asyncio_main())


async def _aiorun_main():
    await corofn()


async def _asyncio_main():
    loop = asyncio.get_event_loop()

    task = loop.create_task(corofn())
    task.add_done_callback(lambda _: loop.stop())

    loop.add_signal_handler(signal.SIGINT, task.cancel)
    loop.add_signal_handler(signal.SIGTERM, task.cancel)

    await task


if __name__ == '__main__':
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s - [%(levelname)s] - [%(name)s] "
               "- %(filename)s:%(lineno)d - %(message)s",
    )

    parser = argparse.ArgumentParser()
    parser.add_argument('--aiorun',
                        action='store_true',
                        help='Run script using aiorun package')
    args = parser.parse_args()

    run(args.aiorun)
```

#### 1) `asyncio.run` + custom signal handlers
```
(.venv) MacBook-Pro-2:~ fedir$ python /Users/fedir/Library/Preferences/PyCharmCE2019.3/scratches/scratch_48.py
^C(.venv) MacBook-Pro-2:~ fedir$
```

#### 2.1) `aiorun.run` + default signal handlers
```
(.venv) MacBook-Pro-2:~ fedir$ python /Users/fedir/Library/Preferences/PyCharmCE2019.3/scratches/scratch_48.py --aiorun 
^C2020-02-06 10:58:13,831 - [CRITICAL] - [aiorun] - aiorun.py:310 - Stopping the loop
2020-02-06 10:58:18,833 - [ERROR] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:283 - Error sending LeaveGroupRequest_v1 to node 1001 [[Error 7] RequestTimedOutError] -- marking coordinator dead
2020-02-06 10:58:18,833 - [ERROR] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:355 - LeaveGroup request failed: [Error 7] RequestTimedOutError
```

#### 2.2) `aiorun.run` + custom signal handlers
And, as you've suggested to override default `aiorun` signal handlers implementation by custom:
```python
async def _aiorun_main():
    loop = asyncio.get_event_loop()
    
    task = asyncio.create_task(corofn())
    task.add_done_callback(lambda _: loop.stop())

    loop.add_signal_handler(signal.SIGINT, task.cancel)
    loop.add_signal_handler(signal.SIGTERM, task.cancel)
    await task
```
no error received when `ERROR` level is set:
```
(.venv) MacBook-Pro-2:~ fedir$ python /Users/fedir/Library/Preferences/PyCharmCE2019.3/scratches/scratch_48.py --aiorun 
^C(.venv) MacBook-Pro-2:~ fedir$ 
```
and proof when `INFO` level is set:
```
[INFO] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:357 - LeaveGroup request succeeded
```
@cjrh Let me ask one more question and then I'm going to close the issue.
Is it the best approach (used in 2.2) to handle such "error on stop phase" case by still using `aiorun` package, what do you think?

**Comment by cjrh** [comment]
@FedirAlifirenko Did you try CTRL-C with `asyncio.run()` _without_ custom signal handlers?

**Comment by cjrh** [comment]
@FedirAlifirenko As to your question - since you can't change aiokafka, you don't have much choice: you will have to control the shutdown sequence manually, for either asyncio.run or with an alternative run implementation like aiorun.  If you are forced to do custom signal handling anyway, maybe you're better off just using the standard library `asyncio.run`, instead of aiorun.

**Comment by FedirAlifirenko** [comment]
> Did you try CTRL-C with asyncio.run() without custom signal handlers?

yes of course, errors are still there:
```python
async def _asyncio_main():
    await corofn()
```
and the output:
```
(.venv) MacBook-Pro-2:~ fedir$ python /Users/fedir/Library/Preferences/PyCharmCE2019.3/scratches/scratch_48.py 
^C2020-02-06 14:44:56,013 - [ERROR] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:283 - Error sending LeaveGroupRequest_v1 to node 1001 [[Error 7] RequestTimedOutError] -- marking coordinator dead
2020-02-06 14:44:56,014 - [ERROR] - [aiokafka.consumer.group_coordinator] - group_coordinator.py:355 - LeaveGroup request failed: [Error 7] RequestTimedOutError
Traceback (most recent call last):
  File "/Users/fedir/Library/Preferences/PyCharmCE2019.3/scratches/scratch_48.py", line 86, in 
...
KeyboardInterrupt
```
but my point is that we haven't any signal handlers at all in this case (whereas with aiorun we have default handlers and we still have the errors)

**Comment by cjrh** [comment]
Ok, I interpret it as "in both cases, custom signal handling is required to deal with aiokafka shutdown".  No worries.  Can we close this issue?

**Comment by FedirAlifirenko** [comment]
@cjrh yes, thanks ✅
